download_images.py

```diff
@@ -0,0 +1,71 @@
+import concurrent.futures
+import itertools
+import json
+import pandas as pd
+import socket
+import urllib.request
+import os
+from os.path import join, exists
+
+class ImageDownloader(object):
+
+    def __init__(self, image_url_df, target_path='data/images', nb_workers=64):
+        self._failed_to_capture_path = 'work/failed_to_capture_images.json'
+        self.image_urls = image_url_df.itertuples()
+        self.target_path = target_path
+        self.nb_workers = nb_workers
+        self.imgs_to_request = image_url_df.shape[0]
+        self.imgs_requested = 0
+        self.failed_img_requests = 0
+        socket.setdefaulttimeout(5)
+
+    def __enter__(self):
+        try:
+            if not exists(self.target_path):
+                os.makedirs(self.target_path)
+            if not exists('work'):
+                os.makedirs('work')
+            with open(self._failed_to_capture_path, 'r') as f:
+                self.failed_to_capture = json.load(f)
+        except:
+            self.failed_to_capture = []
+        return(self)
+
+    def __exit__(self, *args):
+        with open(self._failed_to_capture_path, 'w') as f:
+            json.dump(self.failed_to_capture, f)
+
+    def run(self):
+        print('Images to request: ', self.imgs_to_request)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.nb_workers) as worker_pool:
+            for file_num, fname, img_src in self.image_urls:
+                worker_pool.submit(self.get_one_img(fname, img_src))
+                if (self.imgs_requested % 5000 == 0):
+                    print('Images requested: ', self.imgs_requested)
+                    print('Failed image requests: ', self.failed_img_requests)
+
+    def get_one_img(self, fname, url):
+        local_img_path = join(self.target_path, fname)
+        if exists(local_img_path):
+            return
+        try:
+            self.imgs_requested += 1
+            url, _ = urllib.request.urlretrieve(url, local_img_path)
+        except:
+            import pdb; pdb.set_trace()
+            self.failed_to_capture.append((url, local_img_path))
+            self.failed_img_requests += 1
+
+if __name__ == "__main__":
+    xml_files_by_dir = (i[2] for i in os.walk('data/ILSVRC2014_DET_bbox_train'))
```

Comment:
In these three lines, I just pull the xml file list rather than referencing the bbox file.  That's easy to change. 

Do you see any reason to prefer one of the two options (getting list of images from filesystem vs from your bbox file)?